src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java

```diff
@@ -0,0 +1,127 @@
+  private final InnReachExternalService innReachExternalService;
+  private final RequestService requestService;
+
+  @Transactional
+  @Override
+  public PatronHoldCheckInResponseDTO checkInPatronHoldItem(UUID transactionId, UUID servicePointId) {
+    var transaction = transactionRepository.fetchOneById(transactionId)
+      .orElseThrow(() -> new EntityNotFoundException(String.format("InnReach transaction with id [%s] not found!", transactionId)));
+
+    var hold = (TransactionPatronHold) transaction.getHold();
+    var state = transaction.getState();
+    var shippedItemBarcode = hold.getShippedItemBarcode();
+    var folioItemBarcode = hold.getFolioItemBarcode();
+
+    Assert.isTrue(state == ITEM_SHIPPED, "Unexpected transaction state: " + state);
+    Assert.isTrue(shippedItemBarcode != null, "shippedItemBarcode is not set");
+    Assert.isTrue(folioItemBarcode != null, "folioItemBarcode is not set");
+
+    var checkInResponse = requestService.checkInItem(transaction, servicePointId);
+
+    transaction.setState(ITEM_RECEIVED);
+
+    reportItemReceived(transaction);
+
+    return new PatronHoldCheckInResponseDTO()
+      .transaction(transactionMapper.toDTO(transaction))
+      .folioCheckIn(checkInResponse)
+      .barcodeAugmented(!shippedItemBarcode.equals(folioItemBarcode));
+  }
+
+  @Transactional
+  @Override
+  public ItemHoldCheckOutResponseDTO checkOutItemHoldItem(String itemBarcode, UUID servicePointId) {
+    var transaction = transactionRepository.fetchOneByFolioItemBarcode(itemBarcode)
+      .orElseThrow(() -> new EntityNotFoundException("INN-Reach transaction is not found by itemBarcode: " + itemBarcode));
+
+    var hold = (TransactionItemHold) transaction.getHold();
+    var state = transaction.getState();
+    var folioPatronBarcode = hold.getFolioPatronBarcode();
+
+    Assert.isTrue(state == ITEM_HOLD || state == TRANSFER, "Unexpected transaction state: " + state);
+    Assert.isTrue(folioPatronBarcode != null, "folioPatronBarcode is not set");
+
+    var checkOutResponse = requestService.checkOutItem(transaction, servicePointId);
+    var callNumber = checkOutResponse.getItem().getCallNumber();
+
+    hold.setFolioLoanId(checkOutResponse.getId());
+
+    transaction.setState(ITEM_SHIPPED);
+
+    reportItemShipped(transaction.getTrackingId(), transaction.getCentralServerCode(), itemBarcode, callNumber);
+
+    return new ItemHoldCheckOutResponseDTO()
+      .transaction(transactionMapper.toDTO(transaction))
+      .folioCheckOut(checkOutResponse);
+  }
+
+  private void reportItemReceived(InnReachTransaction transaction) {
+    var centralCode = transaction.getCentralServerCode();
+    var requestPath = resolveItemReceivedPath(transaction.getTrackingId(), centralCode);
+
+    try {
+      innReachExternalService.postInnReachApi(centralCode, requestPath);
+    } catch (InnReachException e) {
+      log.warn("Unexpected D2IR response: {}", e.getMessage(), e);
```

Comment:
Have you thought about reporting the issue back to UI? My point is that the whole action would be completed only partially in this case: everything that belongs to Folio is successful, but Inn-Reach call fails. 
With only logging the error, user will be presented with incomplete result, don't you think?
ps: the same goes to `reportItemShipped` method

Reply:
Right now d2ir API doesn't handle these endpoints appropriately. I assume that the operations may not work as the tracking ids that we are sending are not known by central server

The issue was discussed with PO and for now we agreed to suppress the error https://issues.folio.org/browse/MODINREACH-150?focusedCommentId=116734&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-116734

Reply:
Added a new task to review the flow when d2ir environment is ready for testing these endpoints
https://issues.folio.org/browse/MODINREACH-192